src/main/java/io/cdap/delta/bigquery/MultiGCSWriter.java

```diff
@@ -76,35 +82,71 @@ public synchronized void write(Sequenced<DMLEvent> sequencedEvent) {
     }
   }
 
-  public synchronized Collection<TableBlob> flush() throws IOException {
-    List<TableBlob> writtenObjects = new ArrayList<>(objects.size());
-    IOException error = null;
+  public synchronized Collection<TableBlob> flush() throws IOException, InterruptedException {
+    List<TableBlob> writtenObjects = Collections.synchronizedList(new ArrayList<>(objects.size()));
+    List<Future<?>> writeFutures = new ArrayList<>(objects.size());
     for (Map.Entry<Key, TableObject> entry : objects.entrySet()) {
       TableObject tableObject = entry.getValue();
+      writeFutures.add(executorService.submit((Callable<Void>) () -> {
+        LOG.debug("Writing batch {} of {} events into GCS for table {}.{}", tableObject.batchId, tableObject.numEvents,
+                  tableObject.dataset, tableObject.table);
+        try {
+          tableObject.close();
+        } catch (IOException e) {
+          String errMsg = String.format("Error writing batch of %d changes for %s.%s to GCS",
+                                        tableObject.numEvents, tableObject.dataset, tableObject.table);
+          LOG.error(errMsg, e);
```

Comment:
Don't log if you throw. You can wrap the caught `IOException` with a new message and add the original IOExceptio as the cause.

Reply:
Sure. will wrap and rethrow.